elastic/src/main/java/org/openmbee/sdvc/elastic/services/ElasticSearchService.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Autowired
+    public void setNodeRepository(NodeDAO nodeRepository) {
+        this.nodeRepository = nodeRepository;
+    }
+
+    @Override
+    public ElementsResponse basicSearch(String projectId, String refId, Map<String, String> params) {
+        return recursiveSearch(projectId, refId, params, null);
+    }
+
+    @Override
+    public ElementsResponse recursiveSearch(String projectId, String refId, Map<String, String> params, Map<String, String> recurse) {
+        try {
+            ContextHolder.setContext(projectId, refId);
+            List<Node> validNodes = nodeRepository.findAllByDeleted(false);
+            if(validNodes == null || validNodes.isEmpty()) {
+                return new ElementsResponse();
+            }
+
+            Map<String, ElementJson> elementJsonMap = new HashMap<>();
+            performRecursiveSearch(validNodes, params, recurse, elementJsonMap);
+            return prepareResponse(elementJsonMap.values());
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private void performRecursiveSearch(List<Node> validNodes, Map<String, String> params, Map<String, String> recurse,
+                                        Map<String, ElementJson> elementJsonMap) throws IOException {
+
+        List<ElementJson> elementJsonList = doSearch(validNodes, params);
+        for(ElementJson ob : elementJsonList) {
+            if(! elementJsonMap.containsKey(ob.getId())) {
+                elementJsonMap.put(ob.getId(), ob);
+                Map<String, String> recursiveParams = buildRecursiveParams(ob, recurse);
+                if(! recursiveParams.isEmpty()) {
+                    performRecursiveSearch(validNodes, recursiveParams, recurse, elementJsonMap);
+                }
+            }
+        }
+    }
+
+    private Map<String, String> buildRecursiveParams(ElementJson ob, Map<String, String> recurse) {
+        Map<String, String> recursiveParams = new HashMap<>();
+
+        if(recurse == null || recurse.isEmpty()) {
+            return recursiveParams;
+        }
+
+        for(Map.Entry<String, String> e : recurse.entrySet()) {
+            Object o = ob.get(e.getKey());
+            if(o == null) {
+                continue;
+            }
+            recursiveParams.put(e.getValue(), o.toString());
+        }
+        return recursiveParams;
+    }
+
+
+    private List<ElementJson> doSearch(List<Node> validNodes, Map<String, String> params) throws IOException {
+        List<ElementJson> result = new ArrayList<>();
+
+        SearchRequest searchRequest = new SearchRequest(Index.NODE.get());
+        BoolQueryBuilder query = QueryBuilders.boolQuery();
+        query.filter(QueryBuilders.termsQuery(ElementJson.DOCID, validNodes.stream().map(Node::getDocId).toArray()));
+
+        for(Map.Entry<String, String> e : params.entrySet()) {
+            query.filter(QueryBuilders.termQuery(e.getKey(), e.getValue()));
```

Comment:
fyi when testing, this would only work for things with keyword mapping, anything text or paragraphs would need a `match` query which is outside the scope of this pr